### PR TITLE
feat(chart): v2.0.0 — combined quality improvements (DOPS-434)

### PR DESCRIPTION
**Pull request by pelegvis-BT**
## Summary

Consolidates all DOPS-434 chart improvement work into a single PR (replaces PRs #45-#51).

**Breaking changes:**
- MISSING_CUSTOMER_DATA defaults removed - kmsType now defaults to kubernetes_secret_manager, platform.mode to native
- GCP platform field renamed from platform.gke to platform.gcp (bug fix - values.yaml already used gcp but templates referenced gke)
- imageCredentials and agent.token no longer required in schema (optional with existingSecret)

**New features:**
- agent.existingSecret - reference a pre-existing Secret for GitOps/ESO workflows
- imagePullSecret.existingSecret - reference a pre-existing image pull Secret
- Auto-extraction of imageCredentials from the agent.token blob (no need to pass separately)
- Conditional docker-login Secret creation (skipped when using existingSecret or when no credentials available)
- CI/CD: kind cluster install tests, auto patch version bump on release

**Bug fix:**
- GCP service account annotations were silently empty because _helpers.tpl referenced platform.gke.* while values.yaml defined platform.gcp.*

## Replaces

PRs #45 (DOPS-566), #46 (DOPS-567), #47 (DOPS-568), #48 (DOPS-569), #49 (DOPS-570), #50 (DOPS-571), #51 (DOPS-572)

## Test plan

- [ ] helm lint passes
- [ ] helm template renders correctly for: token path, existingSecret path, GCP platform, AWS platform
- [ ] CI lint-test workflow passes (version bump to 2.0.0 satisfies ct check)
- [ ] Deploy to dev cluster and verify agent starts
- [ ] Verify GCP service account annotation renders correctly with platform.gcp.serviceAccount

Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by guykimchi1 — APPROVED** [review]

File: charts/entitle-agent/ci/integration-test.sh [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+fail() { echo -e "${RED}FAIL${NC}: $1"; FAILED=1; }
+info() { echo -e "${YELLOW}>>>${NC} $1"; }
+
+FAILED=0
+
+# ---------- Helpers ----------
+
+cleanup() {
+  helm uninstall "$RELEASE" -n "$NAMESPACE" 2>/dev/null || true
+  kubectl delete namespace "$NAMESPACE" --wait --timeout=60s 2>/dev/null || true
```
Comment (adi-schreiber) [inline charts/entitle-agent/ci/integration-test.sh]: Drop || true from kubectl delete namespace. If the deletion gets stuck, masking the error will cause confusing failures when the next test tries to recreate it.

File: charts/entitle-agent/ci/integration-test.sh [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    sleep 3
+    elapsed=$((elapsed + 3))
+  done
+
+  echo "  Timed out after ${TIMEOUT}s"
+  kubectl get pods -n "$NAMESPACE" 2>/dev/null || true
+  return 1
+}
+
+check_secret_exists() {
```
Comment (adi-schreiber) [inline charts/entitle-agent/ci/integration-test.sh:130]: The if/else return block is redundant because Bash naturally returns the exit code of the last executed command. You can simplify this entire function to a single line:
check_secret_exists() {
  local name="$1"
  kubectl get secret "$name" -n "$NAMESPACE" >/dev/null 2>&1
}

File: charts/entitle-agent/templates/datadog-secret.yaml [hunk 3]
```diff
@@ -7,7 +9,8 @@ metadata:
   {{- include "entitle-agent.labels" . | nindent 4 }}
 stringData:
 {{- if .Values.datadog.enabled }}
-  api-key: {{ include "entitle-agent.datadogApiKey" . | quote }}
+  api-key: {{ $ddApiKey | quote }}
 {{- else }}
-  DD_API_KEY: {{ include "entitle-agent.datadogApiKey" . | quote }}
+  DD_API_KEY: {{ $ddApiKey | quote }}
```
Comment (adi-schreiber) [inline charts/entitle-agent/templates/datadog-secret.yaml]: Do we need it ?

Comment (pelegvis-BT) [reply charts/entitle-agent/templates/datadog-secret.yaml]: removed

File: charts/entitle-agent/templates/_helpers.tpl [hunk 4]
```diff
@@ -2,27 +2,31 @@
 Expand the name of the chart.
 */}}
 {{- define "entitle-agent.name" -}}
-{{- default "entitle-agent" | trunc 63 | trimSuffix "-" }}
+{{- default "entitle-agent" .Values.nameOverride | trunc 63 | trimSuffix "-" }}
```
Comment (adideutsch) [inline charts/entitle-agent/templates/_helpers.tpl]: @pelegvis-BT we need to properly test all edge cases for changing the name, since we reference this name in several parts of our code (e.g. we run rollout restart entitle-agent, it wont work ootb)

File: charts/entitle-agent/templates/_helpers.tpl [hunk 5]
```diff
@@ -2,27 +2,31 @@
...
 {{/*
 Create a default fully qualified app name.
-We truncate at 63 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
-If release name contains chart name it will be used as a full name.
+Defaults to "entitle-agent" (hardcoded) for backward compatibility with existing
+deployments. Use fullnameOverride to change resource names if needed.
 */}}
 {{- define "entitle-agent.fullname" -}}
-{{- printf "%s" "entitle-agent" | trunc 63}}
+{{- if .Values.fullnameOverride }}
```
Comment (adideutsch) [inline charts/entitle-agent/templates/_helpers.tpl]: @pelegvis-BT we need to properly test all edge cases for changing the name, since we reference this name in several parts of our code (e.g. we run rollout restart entitle-agent, it wont work ootb)

File: charts/entitle-agent/templates/_helpers.tpl [hunk 6]
```diff
@@ -34,43 +38,50 @@ app.kubernetes.io/managed-by: {{ .Release.Service }}
...
 {{- end }}
 
 {{/*
-Service account annotations
+Service account name — returns the configured SA name or the chart fullname.
 */}}
 {{- define "entitle-agent.serviceAccountName" -}}
-{{- default "entitle-agent-sa" | trunc 63 | trimSuffix "-" }}
+{{- if .Values.serviceAccount.create }}
+{{- default (printf "%s-sa" (include "entitle-agent.fullname" .)) .Values.serviceAccount.name | trunc 63 | trimSuffix "-" }}
```
Comment (adideutsch) [inline charts/entitle-agent/templates/_helpers.tpl]: @pelegvis-BT we need to check if the code references the service account name in various validations, and if so make sure it has the right name

File: .github/workflows/lint-test.yaml [hunk 7]
```diff
@@ -1,11 +1,21 @@
+# =============================================================================
```
Comment (adideutsch) [inline .github/workflows/lint-test.yaml:1]: Before releasing this PR, we need to add a documentation change sent to Jenny with the new features so she can add them to our customer-facing docs during the time of release, so users are aware of these new features.

File: charts/entitle-agent/templates/deployment.yaml [hunk 8]
```diff
@@ -19,18 +18,22 @@ spec:
...
       labels:
         app: {{ include "entitle-agent.fullname" . }}
+        {{- include "entitle-agent.selectorLabels" . | nindent 8 }}
         environment: {{ .Values.global.environment }}
         tags.datadoghq.com/env: {{ .Values.global.environment }}
         tags.datadoghq.com/service: {{ include "entitle-agent.fullname" . }}-{{ .Values.global.environment }}
-        tags.datadoghq.com/version: {{ include "entitle-agent.imageTag" .}}
-        {{- include "entitle-agent.selectorLabels" . | nindent 8 }}
+        tags.datadoghq.com/version: {{ include "entitle-agent.imageTag" . }}
+        {{- with .Values.podLabels }}
```
Comment (adideutsch) [inline charts/entitle-agent/templates/deployment.yaml:34]: We also create on demand jobs in customers' clusters,  we need to make sure that those pods also get these pod labels so this feature is consistent across the entire entitle agent setup. (e.g. if they want to use a dedicated node for all of entitle's workloads, omitting these labels from on demand jobs will create a problem for them)

Comment (adideutsch) [reply charts/entitle-agent/templates/deployment.yaml:34]: Also, are these pod labels going to apply to datadog agent damonset pods?

File: charts/entitle-agent/values.yaml [hunk 9]
```diff
@@ -1,70 +1,144 @@
...
+podAnnotations: {}        # https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/
+podLabels: {}             # Additional labels for agent pods
+nodeSelector: {}          # https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#nodeselector
+affinity: {}              # https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#affinity-and-anti-affinity
+tolerations: []           # https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/
+
+# -- kmsType: Key Management Service used to store internal Entitle secrets.
+# Valid values: kubernetes_secret_manager, aws_secret_manager, gcp_secret_manager,
+#               azure_secret_manager, hashicorp_vault
+kmsType: "kubernetes_secret_manager"
```
Comment (adideutsch) [inline charts/entitle-agent/values.yaml:52]: NICE! it's great that we can spare them from this complexity. 99% of users dont change this.

We should probably edit our docs to reflect that this field isnt mandatory anymore.

File: charts/entitle-agent/templates/_helpers.tpl [hunk 10]
```diff
@@ -34,43 +38,50 @@ app.kubernetes.io/managed-by: {{ .Release.Service }}
...
+{{- with .Values.serviceAccount.annotations }}
+{{- toYaml . }}
+{{- end }}
+{{- if eq .Values.platform.mode "aws" }}
 eks.amazonaws.com/role-arn: {{ .Values.platform.aws.iamRole }}
-{{- else if eq .Values.platform.mode "gcp" -}}
-iam.gke.io/gcp-service-account: {{ printf "%s@%s.iam.gserviceaccount.com" .Values.platform.gke.serviceAccount .Values.platform.gke.projectId | quote}}
-{{- else if eq .Values.platform.mode "azure" -}}
+{{- else if eq .Values.platform.mode "gcp" }}
+iam.gke.io/gcp-service-account: {{ printf "%s@%s.iam.gserviceaccount.com" .Values.platform.gcp.serviceAccount .Values.platform.gcp.projectId | quote}}
```
Comment (adideutsch) [inline charts/entitle-agent/templates/_helpers.tpl]: Isn't this a breaking change if someone's values file uses gke?

File: charts/entitle-agent/ci/test-secretref-only.yaml [hunk 11]
```diff
@@ -0,0 +1,32 @@
+# =============================================================================
```
Comment (adideutsch) [inline charts/entitle-agent/ci/test-secretref-only.yaml:1]: @pelegvis-BT is it possible to make the github action runs private , inaccessible to users outside of BT? I don't want people looking at these, prefer to hide it if possible.

File: charts/entitle-agent/templates/_helpers.tpl (hunk 4 above)
Comment (pelegvis-BT) [reply charts/entitle-agent/templates/_helpers.tpl]: will revert if that makes problems

File: charts/entitle-agent/templates/_helpers.tpl (hunk 5 above)
Comment (pelegvis-BT) [reply charts/entitle-agent/templates/_helpers.tpl]: will revert if that make a problem

File: charts/entitle-agent/templates/_helpers.tpl (hunk 6 above)
Comment (pelegvis-BT) [reply charts/entitle-agent/templates/_helpers.tpl]: if the code referes it - i will revert, if not we can keep it as it is a standard helm option

File: .github/workflows/lint-test.yaml (hunk 7 above)
Comment (pelegvis-BT) [reply .github/workflows/lint-test.yaml:1]: created a proposed update for the documentation with the changes: https://beyondtrust.atlassian.net/wiki/x/W4C8yg

File: charts/entitle-agent/values.yaml (hunk 9 above)
Comment (pelegvis-BT) [reply charts/entitle-agent/values.yaml:52]: yep, added it as well to the section that refers to the default kms

File: charts/entitle-agent/templates/_helpers.tpl (hunk 10 above)
Comment (pelegvis-BT) [reply charts/entitle-agent/templates/_helpers.tpl]: you are correct, upgrading the chart and if someone used the gke it can break. the reason behind it was to match the pattern - the platform is the cloud provider and not the k8s service name.
we can revert it

File: charts/entitle-agent/ci/test-secretref-only.yaml (hunk 11 above)
Comment (pelegvis-BT) [reply charts/entitle-agent/ci/test-secretref-only.yaml:1]: will check, although i am not sure it is possible

Comment (pelegvis-BT) [reply charts/entitle-agent/ci/test-secretref-only.yaml:1]: claude response: 
The repo is public. Anyone can:
  - Read the code, workflows, and chart
  - Fork the repo
  - Open PRs from forks

  What outsiders CAN'T do:
  - Push to the repo — only org members with write access can push, so workflows triggered by push only run from
  your team
  - Access secrets — ENTITLE_AGENT_TOKEN is never exposed to fork PRs (GitHub blocks this by design)
  - Run the integration tests — our if condition skips them for fork PRs, and even without it, the secret wouldn't
  be available

  What outsiders CAN do:
  - Open a PR → triggers lint-test job (lint only, no secrets, no cluster). This is harmless.

  To make the repo inaccessible to outsiders, you'd need to make it private. But this is the public chart repo —
  customers need to access it to helm repo add. Making it private would break chart distribution.

  If you want to restrict Actions specifically, two options:
  1. Change the workflow trigger from push: branches: ['**'] to only run on master and your team's branch patterns
  2. Add pull_request_target controls and require approval for first-time contributors

  But honestly, the current setup is already secure — outsiders can only trigger harmless lint, and the token is
  protected.

File: charts/entitle-agent/templates/_helpers.tpl (hunk 4 above)
Comment (pelegvis-BT) [reply charts/entitle-agent/templates/_helpers.tpl]: reverted

File: charts/entitle-agent/templates/_helpers.tpl (hunk 5 above)
Comment (pelegvis-BT) [reply charts/entitle-agent/templates/_helpers.tpl]: reverted

File: charts/entitle-agent/templates/_helpers.tpl (hunk 6 above)
Comment (pelegvis-BT) [reply charts/entitle-agent/templates/_helpers.tpl]: removed

File: .github/workflows/lint-test.yaml [hunk 12]
```diff
@@ -1,11 +1,21 @@
...
+# Lint and Test Charts
+# =============================================================================
+# Runs on every push to any branch:
+#   1. Lint changed charts (yamllint + Yamale schema validation)
+#   2. Smoke test: ct install with minimal values (no real credentials)
+#   3. Integration tests: 3 scenarios with real agent token (requires secret)
```
Comment (adideutsch) [inline .github/workflows/lint-test.yaml:7]: can we completely hide the github action runs from public / external users ?

File: charts/entitle-agent/templates/_helpers.tpl (hunk 10 above)
Comment (pelegvis-BT) [reply charts/entitle-agent/templates/_helpers.tpl]: added the new gcp, and kept the old gke for backward compatibility

File: .github/workflows/lint-test.yaml [hunk 13]
```diff
@@ -42,9 +51,54 @@ jobs:
...
 
-#      - name: Create kind cluster
-#        uses: helm/kind-action@94729529f85113b88f4f819c17ce61382e6d8478 # v1.2.0
-#        if: steps.list-changed.outputs.changed == 'true'
+      # Note: ct install (smoke test) is not used because the agent deployment
+      # has Python-based health probes that require the real agent image.
+      # Real install testing is covered by the integration-test job below.
+
+  # ---------------------------------------------------------------------------
+  # Integration tests — require ENTITLE_AGENT_TOKEN secret
```
Comment (adideutsch) [inline .github/workflows/lint-test.yaml:59]: @pelegvis-BT what environment is the token generated from? prod eu ? other?

Comment (adideutsch) [reply .github/workflows/lint-test.yaml:59]: we need to understand if the token getting revealed by mistake is big deal or not (although its hidden, i'm talking about a potential mistake)

File: .github/workflows/lint-test.yaml (hunk 7 above)
Comment (adideutsch) [reply .github/workflows/lint-test.yaml:1]: we need to review the doc and prepare it for human docs moderator (Jenny?) - and get it approved and released in the same time this PR is released

Comment (adideutsch) [reply .github/workflows/lint-test.yaml:1]: @pelegvis-BT  to make some changes in the doc

File: .github/workflows/lint-test.yaml (hunk 12 above)
Comment (adideutsch) [reply .github/workflows/lint-test.yaml:7]: from discussion with peleg: we cannot, we went over the logs and nothing sensitive is shared. github actions cannot be run by public users because its triggered by PRs and they cant do PRs

File: .github/workflows/lint-test.yaml (hunk 13 above)
Comment (adideutsch) [reply .github/workflows/lint-test.yaml:59]: peleg: it's QA token, saved as gh secret, will not be shown in logs, and if by accident it will ber shown, then it's only to qa, i.e. worst that can happen is spamming S3

File: charts/entitle-agent/ci/test-secretref-only.yaml (hunk 11 above)
Comment (adideutsch) [reply charts/entitle-agent/ci/test-secretref-only.yaml:1]: lets please make sure that outsiders cant open PRs / new branches on the project, it should be read-only for them.

**Review by ZivBleich — APPROVED** [review]

File: charts/entitle-agent/ci/test-secretref-only.yaml (hunk 11 above)
Comment (pelegvis-BT) [reply charts/entitle-agent/ci/test-secretref-only.yaml:1]: outsiders can not open branches / PRs only to fork the repo

File: charts/entitle-agent/templates/deployment.yaml (hunk 8 above)
Comment (pelegvis-BT) [reply charts/entitle-agent/templates/deployment.yaml:34]: These pod labels are added only to deployment, same as you do with values.podAnnotations, both are not used for datadog agent nor the on demand jobs.
@adideutsch Decide if you want to remove that, or add it to datadog agent as well.

Comment (adideutsch) [reply charts/entitle-agent/templates/deployment.yaml:34]: we decided to keep pod labels configurable , and if a customer will use it, and want it to be consistent and apply to jobs as well we will build it for them.
